lamkey_nick_ndvi.py

```diff
@@ -0,0 +1,723 @@
+# 
+# |   | index  | site  | mean_ndvi  | 
+# |---|---|---|---|
+# | Date  |   |   |   |
+# |  2017-01-07  | 0  | SJER  | .4  |  
+# 
+# Be sure to call your dataframe at the end of the cell to ensure autograding works.
+# HINT: FOR THIS STEP, leave any rows containing missing values (`NAN`).
+
+# # Processing multiple scenes from multiple sites 
+# 
+# ### What's the workflow?
+# * Below the same process used above is also used here in the to process multiple sites. The master for loop iterates through two folders. As the master loop iterates a landsat_dir list is generated for a site which is processed in the subsequent for loops the same way as above but now the master for loop will go through both directories containing each site. As long as the landsat imagery is stored in a folder labeled sites with vector layers are named with site_name + -crop.shp, this loop will iterate through any number of sites and scenes. 
+# 
+# ### How was the code optimized?
+# * The loop structure of this code keeps track of each scene so that only the bands from one scene are being worked on through each iteration. Only the date, site names and the final ndvi values are stored outside the loop. The two functions also eliminate variables being stored in memory. Pyproj was used to speed up processing times. 
+
+# In[8]:
+
+
+# Create dataframe of NDVI including the cleaning data to deal with clouds
+
+# Important: to use the ungraded tests below as a sanity check,
+# name your columns: mean_ndvi and site
+# Don't forget to set date as the index and make the values of type datetime
+
+# Path to data
+path_to_data = os.path.join("ndvi-automation", "sites")
+
+# Get list of both site directories
+site_dirs = glob(path_to_data + "/*/")
+
+# Values of clouds from USGS Landsat product guide
+# https://www.usgs.gov/media/files/landsat-8-collection-1-land-surface-reflectance-code-product-guide
+vals = [328, 392, 840, 904, 1350, 352, 368, 416,
+        432, 480, 864, 880, 928, 944, 992, 480, 992]
+
+# Create loop variables
+date_list = []
+site_list = []
+ndvi_list = []
+for i in site_dirs:
+    # Get site names from file path
+    site_names = os.path.basename(os.path.normpath(i))
+    # Set up paths to data in both folders
+    crop_paths = os.path.join(i, "vector")
+    site_crop_paths = os.path.join(crop_paths, site_names + "-crop.shp")
+    crop_bounds = gpd.read_file(site_crop_paths)
+    landsat_dir = os.path.join(i, "landsat-crop")
+    landsat_dirs = sorted(glob(os.path.join(landsat_dir, "LC08*")))
+    # Loop through each directory
+    for k in landsat_dirs:
+
+        # Create dataframe to store data and grab dates and site names
+        mean_ndvi_df = pd.DataFrame()
+        basenames = os.path.basename(k)
+        dates = basenames[10:18]
+        date_list.append(dates)
+        site_list.append(site_names)
+        mean_ndvi_df["date"] = date_list
+        mean_ndvi_df["site"] = site_list
+
+        # Store clean xarrays for masking
+        out_xr = []
+
```

Comment:
@nlamkey  consider removal of some of the "blank"/empty lines in the code - at least within this (github) review environment of the .py file, it is a little difficult for me to follow as my eye is jumping around a bit (the code itself is fine, however).